Addons/tsgExoticQuadrature.hpp

```diff
@@ -0,0 +1,187 @@
+        if (roots[i].size() % 2 == 1) {
+            // Zero out the center for stability.
+            roots[i][(roots[i].size() - 1) / 2] = 0.0;
+        }
+        // Update the values of the polynomials at ref_points.
+        std::swap(poly_m1_vals, poly_vals);
+        std::transform(ref_points.begin(), ref_points.end(), poly_vals.begin(),
+                       [&roots, i](double x){return poly_eval(roots[i], x);});
+    }
+    return roots;
+}
+
+//! \brief Evaluates a Lagrange basis polynomial at a point \b x.
+inline double lagrange_eval(int idx, const std::vector<double> &roots, double x) {
+    double eval_value = 1.0;
+    for (int i=0; i<idx; i++) {
+        eval_value *= (x - roots[i]) / (roots[idx] - roots[i]);
+    }
+    for (size_t i=idx+1; i<roots.size(); i++) {
+        eval_value *= (x - roots[i]) / (roots[idx] - roots[i]);
+    }
+    return eval_value;
+}
+
+//! \brief Generate the full cache of Exotic (weighted and possibly shifted) Gauss-Legendre weights and points for
+//! an (input) max level \b n, using parameters \b shift, \b weight_fn, and \b nref. By default, the reference points
+//! and reference weights are generated by the Gauss-Legendre rule.
+//!
+//! The integrand is assumed to be of the form F(x) = f(x) * weight_fn(x) and the function x->weight_fn(x)+shift
+//! is nonnegative.
+inline TasGrid::CustomTabulated getExoticQuadrature(const int n, const double shift, std::function<double(double)> weight_fn,
+                                                    const int nref, const char* description) {
+
+    // Create the set of reference weights and points for the first term. These reference weights are with respect
+    // to the measure induced by the function [weight_fn(x) + shift].
+    std::vector<double> ref_weights(nref), ref_points(nref);
+    TasGrid::OneDimensionalNodes::getGaussLegendre(nref, ref_weights, ref_points);
+    for (size_t k=0; k<nref; k++) {
+        ref_weights[k] *= (weight_fn(ref_points[k]) + shift);
+    }
+    std::vector<std::vector<double>> points_cache = getRoots(n, ref_weights, ref_points);
+
+    // Create the set of weights for the first term.
+    std::vector<std::vector<double>> weights_cache(n);
+    for (int i=0; i<n; i++) {
+        weights_cache[i] = std::vector<double>(points_cache[i].size(), 0.0);
+        for (size_t j=0; j<points_cache[i].size(); j++) {
+            // Integrate the function l_j(x) according to the measure induced by the function [weight_fn(x) + shift].
+            for (int k=0; k<nref; k++) {
+                weights_cache[i][j] += lagrange_eval(j, points_cache[i], ref_points[k]) * ref_weights[k];
+            }
+        }
+    }
+
+    // Create and append the set of correction points and weights for the second term only if the shift is nonzero.
+    if (shift != 0.0) {
+        for (int i=0; i<n; i++) {
+            const int init_size = points_cache[i].size();
+            std::vector<double>
+                    correction_points(init_size),
```

Comment:
No need to initialize the vectors `getGaussLegendre()` will set the correct size.

Reply:
Changed.